### PR TITLE
MACRO: get rid of `RsIncludeMacroIndex`, use `CrateDefMap` instead

### LINKED ISSUE #9531: Stack overflow on proc macro expansion in included file
## Environment

* **IntelliJ Rust plugin version:** 0.4.181.4937-222-nightly
* **Rust toolchain version:** 1.64.0 (a55dd71d5 2022-09-19) x86_64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2022.2 Ultimate Edition (IU-222.3345.118)
* **Operating system:** macOS 12.5.1
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description

The plugin can constantly throws `StackOverflowError` and even may hang IDE

<details> <summary> Stacktrace </summary>

```
java.lang.StackOverflowError
	at org.rust.lang.core.psi.RsFile.getTreeElement(RsFile.kt:286)
	at com.intellij.psi.impl.source.PsiFileImpl.getStubTree(PsiFileImpl.java:623)
	at com.intellij.psi.impl.source.SpineRef.getStub(SpineRef.java:32)
	at com.intellij.extapi.psi.StubBasedPsiElementBase.getStub(StubBasedPsiElementBase.java:340)
	at com.intellij.psi.util.PsiTreeUtil.getStubChildOfType(PsiTreeUtil.java:435)
	at org.rust.lang.core.psi.impl.RsMetaItemImpl.getPath(RsMetaItemImpl.java:57)
	at org.rust.lang.core.psi.impl.RsMetaItemImpl.getPath(RsMetaItemImpl.java:16)
	at org.rust.lang.core.psi.ext.RsMetaItemKt.getName(RsMetaItem.kt:24)
	at org.rust.lang.core.RsPsiPattern$metaItem$1.invoke(RsPsiPattern.kt:286)
	at org.rust.lang.core.RsPsiPattern$metaItem$1.invoke(RsPsiPattern.kt:285)
	at org.rust.lang.core.RsPsiPatternKt$with$2.accepts(RsPsiPattern.kt:377)
	at com.intellij.patterns.ObjectPattern.accepts(ObjectPattern.java:58)
	at com.intellij.patterns.PatternConditionPlus.process(PatternConditionPlus.java:47)
	at com.intellij.patterns.PatternConditionPlus.process(PatternConditionPlus.java:26)
	at com.intellij.patterns.TreeElementPattern$4.processValues(TreeElementPattern.java:118)
	at com.intellij.patterns.PatternConditionPlus.accepts(PatternConditionPlus.java:42)
	at com.intellij.patterns.ObjectPattern.accepts(ObjectPattern.java:51)
	at com.intellij.patterns.ObjectPattern.accepts(ObjectPattern.java:40)
	at org.rust.lang.core.psi.RsProcMacroPsiUtil.canBeCustomDerive(RsProcMacroPsiUtil.kt:44)
	at org.rust.lang.core.psi.RsProcMacroPsiUtil.canBeProcMacroCall(RsProcMacroPsiUtil.kt:92)
	at org.rust.lang.core.resolve2.FacadeResolveKt.resolveToProcMacroWithoutPsi(FacadeResolve.kt:224)
	at org.rust.lang.core.resolve2.FacadeResolveKt.resolveToProcMacroWithoutPsi$default(FacadeResolve.kt:221)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.resolveToMacroWithoutPsiWithErr(RsPossibleMacroCall.kt:413)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getReasonWhyExpansionFileNotFound-ZVhftvM(MacroExpansionManager.kt:826)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getExpansionFor(MacroExpansionManager.kt:726)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.getExpansionFor(MacroExpansionManager.kt:210)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt._get_expansionResult_$lambda$16(RsPossibleMacroCall.kt:434)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:44)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:121)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.getExpansionResult(RsPossibleMacroCall.kt:428)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.getExpansion(RsPossibleMacroCall.kt:424)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processItem(RsItemsOwner.kt:191)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.access$processItem(RsItemsOwner.kt:1)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt$processItem$1.invoke(RsItemsOwner.kt:176)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt$processItem$1.invoke(RsItemsOwner.kt:175)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processRecursively(RsMacroCall.kt:186)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:180)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processRecursively(RsMacroCall.kt:185)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:180)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:173)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processItem(RsItemsOwner.kt:175)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processExpandedItemsInternal(RsItemsOwner.kt:143)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processExpandedItemsInternal$default(RsItemsOwner.kt:138)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt._get_expandedItemsCached_$lambda$2(RsItemsOwner.kt:66)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.getExpandedItemsCached(RsItemsOwner.kt:62)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsModNullable(FacadeResolve.kt:860)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsMod(FacadeResolve.kt:619)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toScope(FacadeResolve.kt:616)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsMod(FacadeResolve.kt:612)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.findMacroCall-sVsatjs(MacroExpansionManager.kt:777)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.doGetExpandedFromForExpansionFile(MacroExpansionManager.kt:751)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getExpandedFrom$lambda$1(MacroExpansionManager.kt:740)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getExpandedFrom(MacroExpansionManager.kt:738)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.getExpandedFrom(MacroExpansionManager.kt:235)
	at org.rust.lang.core.macros.RsExpandedElementKt.getExpandedFrom(RsExpandedElement.kt:70)
	at org.rust.lang.core.macros.RsExpandedElementKt$findMacroCallExpandedFromNonRecursive$1.invoke(RsExpandedElement.kt:116)
	at org.rust.lang.core.macros.RsExpandedElementKt$findMacroCallExpandedFromNonRecursive$1.invoke(RsExpandedElement.kt:116)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:210)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:170)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:194)
	at kotlin.sequences.SequencesKt___SequencesKt.firstOrNull(_Sequences.kt:167)
	at org.rust.lang.core.macros.RsExpandedElementKt.findMacroCallExpandedFromNonRecursive(RsExpandedElement.kt:117)
	at org.rust.lang.core.macros.RsExpandedElementKt.findMacroCallExpandedFrom(RsExpandedElement.kt:99)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.findIncludingFile(RsMacroCall.kt:137)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.expandIncludeMacroCall(MacroExpansionManager.kt:221)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.getExpansionFor(MacroExpansionManager.kt:209)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt._get_expansionResult_$lambda$16(RsPossibleMacroCall.kt:434)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:121)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.getExpansionResult(RsPossibleMacroCall.kt:428)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.getExpansion(RsPossibleMacroCall.kt:424)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:180)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:173)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processItem(RsItemsOwner.kt:175)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processItem(RsItemsOwner.kt:192)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.access$processItem(RsItemsOwner.kt:1)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt$processItem$1.invoke(RsItemsOwner.kt:176)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt$processItem$1.invoke(RsItemsOwner.kt:175)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processRecursively(RsMacroCall.kt:186)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:180)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processRecursively(RsMacroCall.kt:185)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:180)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.processExpansionRecursively(RsMacroCall.kt:173)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processItem(RsItemsOwner.kt:175)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processExpandedItemsInternal(RsItemsOwner.kt:143)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.processExpandedItemsInternal$default(RsItemsOwner.kt:138)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt._get_expandedItemsCached_$lambda$2(RsItemsOwner.kt:66)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.psi.ext.RsItemsOwnerKt.getExpandedItemsCached(RsItemsOwner.kt:62)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsModNullable(FacadeResolve.kt:860)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsMod(FacadeResolve.kt:619)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toScope(FacadeResolve.kt:616)
	at org.rust.lang.core.resolve2.FacadeResolveKt.toRsMod(FacadeResolve.kt:612)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.findMacroCall-sVsatjs(MacroExpansionManager.kt:777)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.doGetExpandedFromForExpansionFile(MacroExpansionManager.kt:751)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getExpandedFrom$lambda$1(MacroExpansionManager.kt:740)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.macros.MacroExpansionServiceImplInner.getExpandedFrom(MacroExpansionManager.kt:738)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.getExpandedFrom(MacroExpansionManager.kt:235)
	at org.rust.lang.core.macros.RsExpandedElementKt.getExpandedFrom(RsExpandedElement.kt:70)
	at org.rust.lang.core.macros.RsExpandedElementKt$findMacroCallExpandedFromNonRecursive$1.invoke(RsExpandedElement.kt:116)
	at org.rust.lang.core.macros.RsExpandedElementKt$findMacroCallExpandedFromNonRecursive$1.invoke(RsExpandedElement.kt:116)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:210)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:170)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:194)
	at kotlin.sequences.SequencesKt___SequencesKt.firstOrNull(_Sequences.kt:167)
	at org.rust.lang.core.macros.RsExpandedElementKt.findMacroCallExpandedFromNonRecursive(RsExpandedElement.kt:117)
	at org.rust.lang.core.macros.RsExpandedElementKt.findMacroCallExpandedFrom(RsExpandedElement.kt:99)
	at org.rust.lang.core.psi.ext.RsMacroCallKt.findIncludingFile(RsMacroCall.kt:137)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.expandIncludeMacroCall(MacroExpansionManager.kt:221)
	at org.rust.lang.core.macros.MacroExpansionManagerImpl.getExpansionFor(MacroExpansionManager.kt:209)
	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt._get_expansionResult_$lambda$16(RsPossibleMacroCall.kt:434)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:230)
```

</details>

The exception doesn't appear if procedural macro expansion is disabled

Created from https://github.com/intellij-rust/intellij-rust/issues/9402#issuecomment-1276067045

## Steps to reproduce

- open https://github.com/drcarney-sp/ij-issue/tree/81ea82713c1e7526c713cdd8afec57b8b4da0db2
- uncomment https://github.com/drcarney-sp/ij-issue/blob/81ea82713c1e7526c713cdd8afec57b8b4da0db2/src/lib.rs#L15

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #9531

changelog: Improve `include!` macros search

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt [hunk 1]
```diff
@@ -776,15 +802,34 @@ private class MacroExpansionServiceImplInner(
...
+                    return@run getExpansionFile(defMap, parentIndex)
+                }
+                nestedIndices += parentIndex.last
+            }
+            @Suppress("UNREACHABLE_CODE")
+            null
         } ?: return null
+
+        var parent: RsMod = nearestKnownParent
+        for (macroIndexInParent in nestedIndices) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt:827]: Shouldn't we iterate `nestedIndices` in reverse order here? If yes lets add a test for it

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt]: Adjusted. This case should now be covered by `test include file in included file in included file`

**Review by dima74 — COMMENTED** [review]
Maybe extract first commit into separate PR?

I ran [extended regression testing](https://github.com/dima74/intellij-rust/actions/runs/3519121147) and it found assertion failure in file `petgraph-0.5.1/src/algo/mod.rs`

```
org.rustPerformanceTests.CustomRealProjectAnalysisTest > test[rust-analyzer] STANDARD_ERROR
    ERROR: macro.expansion.context != macro.context; macro: `RsMacroCallImpl(MACRO_CALL)`, expandedElementContext: `RsConstantImpl(CONSTANT)`
    java.lang.IllegalStateException: macro.expansion.context != macro.context; macro: `RsMacroCallImpl(MACRO_CALL)`, expandedElementContext: `RsConstantImpl(CONSTANT)`
    	at org.rust.lang.core.psi.ext.RsPossibleMacroCallKt.checkExpansionResult(RsPossibleMacroCall.kt:448)
```

**Comment by vlad20012** [comment]
> Maybe extract first commit into separate PR?

I can extract it only without the assert because the assert fails on some `include` macros without the second commit. And I don't think the commit without that assert is very useful since you can't check it using regression tests for example

**Review by dima74 — APPROVED** [review]
bors r+

**Pull request merged**